[AST] Store SubstitutionMaps in the AST rather than Substitution lists.

DougGregor opened the pull request:
Consistently store `SubstitutionMap` within various parts of the AST rather than `Substitution` lists, and deal with the fallout. Specific changes:

* `ConcreteDeclRef` now uses `SubstitutionMap` and has been simplified.
* `Witness` now uses `SubstitutionMap`
* `SpecializedProtocolConformance` now uses `SubstitutionMap`
* `SILBoxType` now uses `SubstitutionMap`.

Overall, removes ~90 explicit uses of `SubstitutionList` (of ~410).

DougGregor committed:
[AST] Store SubstitutionMaps in ConcreteDeclRef and Witness data structures.

Replace two prominent uses of SubstitutionList, in ConcreteDeclRef and
Witness, with SubstitutionMap. Deal with the myriad places where we
now have substitution maps and need substitution lists (or vice versa)
caused by this change.

Overall, removes ~50 explicit uses of SubstitutionList (of ~400).

slavapestov commented:
Yay!

slavapestov commented on include/swift/SIL/SILFunction.h:
```diff
@@ -123,6 +123,9 @@ class SILFunction
   Optional<SubstitutionList> ForwardingSubs;
 
+  /// The forwarding substitution map, lazily computed.
+  SubstitutionMap ForwardingSubMap;
```
Perhaps we should move this caching logic to GenericEnvironment?

DougGregor replied:
So little code cares about it that it doesn't seem necessary.

slavapestov commented on lib/AST/ASTVerifier.cpp:
```diff
@@ -2416,6 +2416,10 @@ class Verifier : public ASTWalker {
     }
 
+    void verifyChecked(SubstitutionMap substitutions){
+      // FIXME: Check replacement types without forcing anything.
```
SubstitutionMap has a verify() method

slavapestov commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -4805,11 +4812,19 @@ emitSpecializedAccessorFunctionRef(SILGenFunction &SGF,
+  SubstitutionMap subMap;
+  if (auto genericFnType = dyn_cast<GenericFunctionType>(origAccessType)) {
+    auto genericSig = genericFnType.getGenericSignature();
+    subMap = genericSig->getSubstitutionMap(substitutions);;
```
Unnecessary ;, and this could use toList() instead

DougGregor replied:
It's going the other way, `SubstitutionList` -> `SubstitutionMap`. I'll remove the extra `;` of course.

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3371,27 +3366,44 @@ static void checkVarBehavior(VarDecl *decl, TypeChecker &TC) {
+  // Build interface and context substitution maps.
+  auto protoSelfType = behaviorProto->getProtocolSelfType();
+  auto sig = behaviorProto->getGenericSignatureOfContext();
+  auto interfaceSubsMap =
```
This is SubstitutionMap::getProtocolSubstitutions()

DougGregor replied:
Oh, of course it is! Thanks.

slavapestov reviewed (commented):
I have a few suggestions after taking another look :-)

DougGregor committed:
[AST] Switch SpecializedProtocolConformance over to SubstitutionMap.

Eliminate another common use of SubstitutionList from the AST.

DougGregor committed:
[AST] Use SubstitutionMap, not SubstitutionList, in SILBoxType.

Eliminate the last place in the AST proper that stores
SubstitutionLists rather than SubstitutionMaps. Randomly fixes a
crasher, too.

DougGregor committed:
[AST] Collapse storage for ConcreteDeclRef.

Rather than ASTContext-allocating ConcreteDeclRef’s storage when there is a
non-empty substitution map, put the SubstitutionMap directly in the
ConcreteDeclRef. Simplify the various interfaces along the way.

DougGregor committed:
[SILGen] Use the SubstitutionMap form of emitApplyOfLibraryIntrinsic().

DougGregor commented:
And... had to rebase. Comments addressed, `SpecializedProtocolConformance` and `SILBoxType` converted. `ConcreteDeclRef` flattened.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
@swift-ci please test compiler performance
